src/utility.ts

```diff
@@ -248,3 +250,24 @@ export async function waitForStandardMode(): Promise<boolean> {
 
     return true;
 }
+
+export async function searchMainMethods(...uris: vscode.Uri[] | undefined): Promise<IMainClassOption[]> {
+    try {
+        return await vscode.window.withProgress<IMainClassOption[]>(
+            { location: vscode.ProgressLocation.Window },
+            async (p) => {
+                p.report({ message: "Searching main classes..." });
+                if (_.isEmpty(uris)) {
+                    uris = vscode.workspace.workspaceFolders.map((folder) => folder.uri);
+                }
+                const mainClasses: IMainClassOption[] = [];
+                for (const uri of uris) {
+                    mainClasses.push(...await resolveMainClass(uri));

```

Comment:
Calling search engine multiple times is not efficient. In current PR, i didn't see there is requirement to search for multiple uris at one time. I suggest to keep passing at most 1 uri in this method. If there is requirement in future to search for multiple uris, we can improve the Java side logic for that.

Reply:
But what if the workspace has multiple folders? calling `searchMainMethods()` multiple times?

Reply:
If you didn't pass uri in this method, it means searching in all folders.

If you look at the Java implementation, what it did is searching all results first. and if there is a uri given, then filter the result with the uri, otherwise, return all.

Reply:
https://github.com/microsoft/java-debug/blob/master/com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java#L77

Reply:
I see